// @storybook/svelte/dist/cjs/client/preview/render.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = render;

require("core-js/modules/es.function.name.js");

var _global = _interopRequireDefault(require("global"));

var _PreviewRender = _interopRequireDefault(require("./PreviewRender.svelte"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var document = _global.default.document;
var previousComponent = null;

function cleanUpPreviousStory() {
  if (!previousComponent) {
    return;
  }

  previousComponent.$destroy();
  previousComponent = null;
}

function render(_ref) {
  var storyFn = _ref.storyFn,
      kind = _ref.kind,
      name = _ref.name,
      showMain = _ref.showMain,
      showError = _ref.showError;
  cleanUpPreviousStory();
  var target = document.getElementById('root');
  target.innerHTML = '';
  previousComponent = new _PreviewRender.default({
    target: target,
    props: {
      storyFn: storyFn,
      name: name,
      kind: kind,
      showError: showError
    }
  });
  showMain();
}
